File: server/TWS Foundation/CSM Foundation/Database/Entity/Models/EntityOperationFailure.cs
using System.Text.Json.Serialization;

namespace CSM_Foundation.Database.Entity.Models;


/// <summary>
///     [Record] that stores the information about a failed operation over an <see cref="IEntity"/>.
/// </summary>
/// <typeparam name="TEntity">
///     Type of the <see cref="IEntity"/> that failed.
/// </typeparam>
public record EntityOperationFailure<TEntity>
    where TEntity : IEntity {

    /// <summary>
    ///     Original instance that the operation was committed and failed.
    /// </summary>
    public TEntity Entity { get; init; } = default!;

    /// <summary>
    ///     Message caugth from the immediate exception that caused the operation failure.
    /// </summary>
    public string Message { get; init; } = string.Empty;

    /// <summary>
    ///     <see cref="System.Exception"/> caught over the operation failure, the information to trace
    ///     all about the operation failure at system level.
    /// </summary>
    [JsonIgnore]
    public Exception Exception { get; init; } = default!;

    /// <summary>
    ///     Creates a new <see cref="EntityOperationFailure{TEntity}"/> instance.
    /// </summary>
    /// <param name="entity">
    ///     <typeparamref name="TEntity"/> instance targeted by the operation that failed.
    /// </param>
    /// <param name="exception">
    ///     <see cref="System.Exception"/> caught over the operation failure.
    /// </param>
    public EntityOperationFailure(TEntity entity, Exception exception) {
        Entity = entity;
        Exception = exception;
        Message = exception.Message;
    }
}

File: client/tws_foundation_client/lib/src/services/models/entity_operation_failure.dart
import 'package:csm_client_core/csm_client_core.dart';
import 'package:tws_foundation_client/tws_foundation_client.dart';

/// {implementation} class for a [EntityOperationFailure].
///
/// Defines an object that stores information about an [IEntity] operation failure result at a [FoundationServer] service request.
final class EntityOperationFailure<T extends IEntity<T>> implements IDecodable, IEncodable {
  /// Key binding for [entity] property..
  static const String kEntity = 'entity';

  /// Key binding for [message] property.
  static const String kMessage = 'message';

  /// Original instance that the operation was committed and failed..
  late T entity;

  /// Message caugth from the immediate exception that caused the operation failure.
  String message = '';

  /// Object factory for [T] types.
  final T Function() entityFactory;

  /// Creates a new [EntityOperationFailure] instance.
  EntityOperationFailure(this.entityFactory);

  @override
  void decode(DataMap encode) {
    message = encode.get(kMessage);

    entity = entityFactory();
    entity.decode(encode.get(kEntity));
  }

  @override
  DataMap encode() {
    return <String, Object?>{
      'entity': entity.encode(),
      'message': message,
    };
  }
}
